qmanager: add multi-queue support

dongahn opened the pull request:
This PR is work in progress to add multi-queue support within `qmanager`.

Build on top of a merge of two pending PRs: PR #645 and PR #646.

As described in Issue #640, it adds the following modifications:
 - Replace a single queue object (of `queue_policy_base_t` class) with
a `std::map` that contains a configurable number of queue objects
keyed off of distinct queue names.
- Add ways to specify multiple queues both through configuration files
and module load options while keeping the current conf/load option
set backward compatible as much as possible. 
Each specified queue can enforce a queueing policy, queuing
parameters and queueing-policy parameters, independent of one another.
- Bring in libjobspec to fetch the queue information from
`system.attributes.queue` key of a jobspec and use
it to identify the corresponding queue.
- Adjust the business logic in each and every job-manager callbacks
like `jobmanager_alloc_cb`
- Add a few configuration tests into `t1005-qmanager-conf.t`

I still need to add a bunch more test cases to finish this up.
As we discussed before, I'm not entirely happy with the current
way of configuring  queues. But I thought there is a benefit of getting
the runtime logic correct as soon as possible and then improve
upon user friendliness later.

If you want to test this:

```console
ahn1@6e8124a3b9bc:/usr/src$ flux start -s 1
ahn1@6e8124a3b9bc:/usr/src$ flux module remove qmanager
ahn1@6e8124a3b9bc:/usr/src$ flux module load qmanager queues="debug batch"
ahn1@6e8124a3b9bc:/usr/src/t/multiqueue$ flux mini run -n 2 --dry-run hostname | jq '.attributes.system.queue="batch"' | flux job submit -
10022792724480
ahn1@6e8124a3b9bc:/usr/src/t/multiqueue$ flux mini run -n 2 --dry-run hostname | jq '.attributes.system.queue="debug"' | flux job submit -
10185330393088
ahn1@6e8124a3b9bc:/usr/src/t/multiqueue$ flux mini run -n 2 --dry-run hostname | flux job submit -
10675594199040
ahn1@6e8124a3b9bc:/usr/src/t/multiqueue$ flux dmesg | grep alloc
2020-05-02T00:38:41.891482Z qmanager.debug[0]: alloc success (queue=batch id=10022792724480)
2020-05-02T00:38:51.581285Z qmanager.debug[0]: alloc success (queue=debug id=10185330393088)
2020-05-02T00:39:20.800212Z qmanager.debug[0]: alloc success (queue=batch id=10675594199040)
```

grondo commented:
> ahn1@6e8124a3b9bc:/usr/src/t/multiqueue$ flux mini run -n 2 --dry-run hostname | jq '.attributes.system.queue="batch"' | flux job submit -
10022792724480

Just an FYI - this can more easily be done via:

```
$ flux mini submit -n2 --setattr system.queue=batch hostname
```

dongahn commented:
> $ flux mini submit -n2 --setattr system.queue=batch hostname

Yeah @SteVwonder and I discussed this as the way to enqueue a job into a non-default queue way back. But I completely forgot and used a much longer way to express it :-). 

Nice job on the flux-mini interface!

dongahn commented:
Regarding the config file specification form, please see https://github.com/flux-framework/flux-sched/issues/650

dongahn commented:
> ahn1@6e8124a3b9bc:/usr/src/t/multiqueue$ flux dmesg | grep alloc
2020-05-02T00:38:41.891482Z qmanager.debug[0]: alloc success (queue=batch id=10022792724480)
2020-05-02T00:38:51.581285Z qmanager.debug[0]: alloc success (queue=debug id=10185330393088)
2020-05-02T00:39:20.800212Z qmanager.debug[0]: alloc success (queue=batch id=10675594199040)

I need to add test cases to finish this up and pass/failure will be whether the job has been enqueued into the named queue correctly.

I wonder how I can validate this when we don't have a way to display scheduler info like queues through job listing tool like `flux jobs`, 

My expectation is some of the key scheduler info will ultimately be displayed using `flux jobs` by utilizing the "note" field -- https://github.com/flux-framework/flux-sched/issues/626

@chu11 or @garlick: do you know when we will able to add note support in job listing tools within flux-core?

dongahn commented:
Just a note: when PR #645 and #646 are merged into master, a merge conflict will need to be resolved.

dongahn commented:
OK. I have added some more test cases and split larger commits into smaller atomic commits. This is ready to be reviewed. Once this lands, I will do name changing PR per Issue #485.

SteVwonder commented on qmanager/modules/qmanager_opts.cpp:
```diff
@@ -204,6 +204,15 @@ qmanager_opts_t &qmanager_opts_t::operator+= (const qmanager_opts_t &src)
+bool qmanager_opts_t::operator ()(const std::string &k1,
+                                  const std::string &k2) const
+{
+    if (m_tab.find (k1) == m_tab.end ()
+        || m_tab.find (k2) == m_tab.end ())
+        return k1 < k2;
```
If one of the keys is in `m_tab` and the other is not, would you want to handle the one that is in `m_tab` before handling the other?

dongahn replied:
I think in this case this would be okay as the unknown keys will be just ignored. What we need to make sure is the known keys get to process in order. In the case of the `qmanager_opts_t` class, only the `queues=` option is required to be processed first.

SteVwonder commented on qmanager/modules/qmanager_opts.cpp:
```diff
@@ -79,94 +163,236 @@ qmanager_opts_t::qmanager_opts_t ()
+            throw std::bad_alloc ();
+    }
+    for (auto &kv : m_per_queue_prop) {
+        if (!is_default_queue_set ())
+            m_default_queue = kv.first;
```
Will this ever be true due to the setting of `m_default_queue` to the first entry in `parse_queues`?

dongahn replied:
Yes, this will be true for example if `queues=qbatch qdebug` is given. In that case, the `m_default_queue` would be `0xdeadbeef` and set to the named queue that is lexicographically smallest. On the other hand, if `default-queue=qdebug` is passed also, `m_default_queue` is set to that so the default queue becomes `qdebug` instead of `qbatch.`

If this is not the case, there is a bug somewhere :-)

SteVwonder replied:
> and set to the named queue that is lexicographically smallest.

I guess what I'm getting at is that wouldn't that setting be done in the [`parse_queues` method](https://github.com/flux-framework/flux-sched/pull/649/files/db42eff456a9beb040cfce008b5b236d3bbd7731#diff-ecff1f5a8c8cc615ce9f7c7ca69362cfR47) causing this particular code to never run because `is_default_queue_set` will always return true?  For example, I applied the patch pasted below and could not seem to trigger this code to run, including with the example you provided.

Patch:
```diff
diff --git a/qmanager/modules/qmanager_opts.cpp b/qmanager/modules/qmanager_opts.cpp
index aaf7fdb2..7557f7e1 100644
--- a/qmanager/modules/qmanager_opts.cpp
+++ b/qmanager/modules/qmanager_opts.cpp
@@ -22,6 +22,7 @@
  *  See also:  http://www.gnu.org/licenses/
 \*****************************************************************************/

+#include <iostream>
 #include "qmanager_opts.hpp"

 using namespace Flux;
@@ -43,8 +44,10 @@ int qmanager_opts_t::parse_queues (const std::string &queues)
             goto done;
         m_per_queue_prop.clear (); // clear the default queue entry
         for (const auto &entry: entries) {
-            if (m_default_queue == m_default_queue_name)
+            if (m_default_queue == m_default_queue_name) {
+                std::cerr << "Parse_queues setting m_default_queue to " << entry << std::endl;
                 m_default_queue = entry;
+            }
             auto ret = m_per_queue_prop.insert (
                            std::pair<std::string, queue_prop_t> (
                                entry, queue_prop_t ()));
@@ -268,8 +271,10 @@ qmanager_opts_t &qmanager_opts_t::canonicalize ()
             throw std::bad_alloc ();
     }
     for (auto &kv : m_per_queue_prop) {
-        if (!is_default_queue_set ())
+        if (!is_default_queue_set ()) {
+            std::cerr << "Canonicalize setting m_default_queue to " << kv.first << std::endl;
             m_default_queue = kv.first;
+        }
         if (!kv.second.is_queue_policy_set ())
             kv.second.set_queue_policy (m_queue_prop.get_queue_policy ());
         if (!kv.second.is_queue_params_set ())
```

Example:
```
sherbein .../hdd/Repositories/flux-framework/flux-sched (multi-queue !?S)
❯ flux module reload qmanager queues="debug batch"                                                                                                                                                                                                                   23:58:38 ()
Canonicalize setting m_default_queue to batch

sherbein .../hdd/Repositories/flux-framework/flux-sched (multi-queue !?S)
❯ flux module reload qmanager queues="debug batch" default-queue="batch"                                                                                                                                                                                             23:58:49 ()

sherbein .../hdd/Repositories/flux-framework/flux-sched (multi-queue !?S)
❯ flux module reload qmanager queues="debug batch" default-queue="debug"                                                                                                                                                                                             23:59:02 ()

sherbein .../hdd/Repositories/flux-framework/flux-sched (multi-queue !?S)
❯ flux module reload qmanager                                                                                                                                                                                                                                        23:59:06 ()
Canonicalize setting m_default_queue to default
```

dongahn replied:
Ah. Sorry I misunderstood your comment. Yes, I think this is a dead code, that remained after a code refactoring. Will fix.

dongahn replied:
Hmmmm. I will have to take it back. Some tests are failing without this code. Let me look more.

dongahn replied:
My apology. I don't think I understand your comment any longer. You are not saying Line 272 will never be executed, are you? Your final example 

> sherbein .../hdd/Repositories/flux-framework/flux-sched (multi-queue !?S)
❯ flux module reload qmanager                                                                                                                                                                                                                                        23:59:06 ()
Canonicalize setting m_default_queue to default

shows it is being set to "default"

Maybe you are referring to Line 47 within `parse_queues` instead?

dongahn replied:
> Maybe you are referring to Line 47 within parse_queues instead?

Ok. I think this is a dead code. I will eliminate that.

dongahn commented:
> Thanks @dongahn for putting this together! Little stuff is inline. Generally LGTM!

> The only big thing is that it is possible for the handshake with the job-manager to request single mode despite the user setting a queue policy of easy. After applying a patch to log the job-manager queue mode, I tried loading qmanager with the queue-policy-per-queue and got a surprising result of single. It looks like the handshake function is only checking the main queue property object for the policy and setting the job-manager queue mode based on that. If you don't specify queue-policy, but you do specify queue-policy-per-queue, you can get an incorrect state.

Ah... This is a great catch! 

With multi queue support, I am not even sure if "single" mode even makes sense any longer. If you have two FCFS queues at `qmanager`, you don't want "single" from `job-manager` reducing job throughput... The only mode where "single" makes sense is when you have a single queue within `qmanager` and it enforces FCFS. 

If there isn't much performance difference between `single` and `unlimited` for this case, we may want to deprecate "single" at `qmanager` for good...

dongahn commented on qmanager/modules/qmanager_opts.hpp:
```diff
@@ -95,10 +136,12 @@ class qmanager_opts_t : public optmgr_parse_t {
+    const std::string m_default_queue_name = "default";
+    std::string m_default_queue = QMANAGER_OPTS_UNSET_STR;
     queue_prop_t m_queue_prop;
+    std::map<std::string, queue_prop_t> m_per_queue_prop;
     std::map<std::string, int> m_tab;
```
This is used to quite often to indicate the variables are C++ class member variables. https://stackoverflow.com/questions/13018189/what-does-m-variable-prefix-mean.

Yes I can add some notes about these variables.

dongahn commented on qmanager/modules/qmanager_opts.cpp:
```diff
@@ -204,6 +204,15 @@ qmanager_opts_t &qmanager_opts_t::operator+= (const qmanager_opts_t &src)
     return canonicalize ();
 }
 
+bool qmanager_opts_t::operator ()(const std::string &k1,
```
Good catch. Will fix.

dongahn commented on qmanager/modules/qmanager_opts.cpp:
```diff
@@ -79,94 +163,236 @@ qmanager_opts_t::qmanager_opts_t ()
+        if (!m_per_queue_prop.empty ()
+            && m_per_queue_prop.find (v) == m_per_queue_prop.end ()) {
+            info += "Unknown default queue (" + v + ")! ";
+            info += "Use default.";
```
Agreed. Will change.

dongahn commented on qmanager/modules/qmanager_opts.cpp:
```diff
@@ -79,94 +163,236 @@ qmanager_opts_t::qmanager_opts_t ()
-        if (!known_queue_policy (m_queue_prop.queue_policy)) {
+        if (!m_queue_prop.set_queue_policy (v)) {
             info += "Unknown queuing policy (" + v + ")! ";
             info += "Use default.";
```
Ditto.

dongahn commented on t/t1006-qmanager-multiqueue.t:
```diff
@@ -0,0 +1,152 @@
+#!/bin/sh
```
Will change. Feels like I can benefit from a spell checker for git commit messages.

SteVwonder commented:
> If there isn't much performance difference between single and unlimited for this case, we may want to deprecate "single" at qmanager for good...

Sounds good to me.

dongahn commented:
I will work on it today to land it.

dongahn commented:
> If there isn't much performance difference between single and unlimited for this case, we may want to deprecate "single" at qmanager for good...

Looks to me like `qmanage` performs about the same whether it is `unlimited` or `single`. I will just use `unlimited`.

unlimited, fcfs
```console
ahn1@90127f184db7:/usr/src/t/perf_test$ ./bulksubmit.py jobs/*.json
bulksubmit: Starting...
bulksubmit: submitted 1024 jobs in 3.98s. 257.25job/s
bulksubmit: First job finished in about 3.983s

bulksubmit: Ran 1024 jobs in 61.4s. 16.7 job/s
```

single, fcfs
```console
ahn1@90127f184db7:/usr/src/t/perf_test$ ./bulksubmit.py jobs/*.json
bulksubmit: Starting...
bulksubmit: submitted 1024 jobs in 3.76s. 272.30job/s
bulksubmit: First job finished in about 3.763s

bulksubmit: Ran 1024 jobs in 64.3s. 15.9 job/s
```

dongahn commented:
@SteVwonder: Ok. I addressed all of the issues above. If Travis CI turns green, this should be good to go.

dongahn commented:
All green now!
